projects/l2_pooling/multi_column_convergence.py

```diff
@@ -38,54 +40,91 @@
   createObjectMachine
 )
 
-def locateConvergencePoint(stats, targetValue):
+def locateConvergencePoint(stats, minOverlap, maxOverlap):
   """
   Walk backwards through stats until you locate the first point that diverges
-  from targetValue.  We need this to handle cases where it might get to
-  targetValue, diverge, and then get back again.  We want the last convergence
-  point.
+  from target overlap values.  We need this to handle cases where it might get
+  to target values, diverge, and then get back again.  We want the last
+  convergence point.
   """
   for i,v in enumerate(stats[::-1]):
-    if v != targetValue:
-      return len(stats)-i
+    if not (v >= minOverlap and v <= maxOverlap):
+      return len(stats)-i + 1
 
-  # Never differs - converged right away
-  return 0
+  # Never differs - converged in one iteration
+  return 1
 
 
-def averageConvergencePoint(inferenceStats, prefix, targetValue):
+def averageConvergencePoint(inferenceStats, prefix, minOverlap, maxOverlap,
+                            settlingTime):
   """
+  inferenceStats contains activity traces while the system visits each object.
+
+  Given the i'th object, inferenceStats[i] contains activity statistics for
+  each column for each region for the entire sequence of sensations.
+
+  For each object, compute the convergence time - the first point when all
+  L2 columns have converged.
+
+  Return the average convergence time across all objects.
+
   Given inference statistics for a bunch of runs, locate all traces with the
   given prefix. For each trace locate the iteration where it finally settles
   on targetValue. Return the average settling iteration across all runs.
   """
-  itSum = 0
-  itNum = 0
+  convergenceSum = 0.0
+
+  # For each object
   for stats in inferenceStats:
+
+    # For each L2 column locate convergence time
+    convergencePoint = 0.0
     for key in stats.iterkeys():
       if prefix in key:
-        itSum += locateConvergencePoint(stats[key], targetValue)
-        itNum += 1
+        convergencePoint = max(
+          convergencePoint,
+          locateConvergencePoint(stats[key], minOverlap, maxOverlap))
```

Comment:
Independent of whether it's a bug, should we let this script be fooled? I don't think it's super-complicated to find the first touch where every column has ~40 active cells, though the "inferenceStats" approach does seem to complicate things. (rather than just looping over `exp.infer()` and doing the touch counting yourself, and checking exp.getL2Representations() after every touch)

If this is a bug, it's not a bug in the code. Whether a cell is active is determined completely from the feedforward and lateral input (and not from the cell's previous state), except when there's little-to-no feedforward input. If L2 columns could lock in on an object and continue clutching onto it even when it receives more evidence for other objects laterally, then it would be error-prone in situations where you switch the object or when there's noise. You'd get silly situations where 4 columns infer Object B while 1 column stays locked in on Object A. In the current model, the cell's previous state contributes 1 active segment to the lateral support, but it can still be outvoted by other segments.

In the example above, every Object A cell always has 4 active segments while every Object B cell has 3, and 4 segments is not enough to inhibit 3.